### ISSUE TITLE
Allow more powerful readdir

### ISSUE THREAD

**Issue by netheril96** [billziss-gh/winfsp#44]
The FUSE API `readdir` only reads the file mode from `stat` buffer passed to filler. I guess that is why in WinFsp each `readdir` call is accompanied by many `getattr` since the NT API returns both name and attributes during directory traversing. This has proved to be a major performance bottleneck in my application, and I would like to see an option to let WinFsp know that the `stat` buffer I passed to `fuse_filler_t` is a fully initialized `stat` structure, so the additional `getattr` calls will be eliminated.

**Comment by billziss-gh** [comment]
This is a great suggestion and it is something that has bothered me as well. Unfortunately the FUSE 2.8 API that WinFsp implements does not support the `FUSE_READDIR_PLUS` flag that became available in FUSE 3.

Perhaps one way to shoehorn this on is to introduce an `FSP_FUSE_READDIR_PLUS` conditional macro that defines `fuse_readdir_flags`, `fuse_fill_dir_flags` and changes the definitions of `fuse_fill_dir_t` and `readdir` to match those of FUSE 3.0. We can consider this as an optional WinFsp extension to FUSE 2.8.

Unfortunately this is not a trivial change. The primary problem is the `flags` parameter to `fuse_fill_dir_t`.

https://github.com/libfuse/libfuse/blob/fuse_3_0_bugfix/include/fuse.h#L76

While one can "safely" pass extra parameters to a (`cdecl`) function in C, the actual implementation of a function must know whether the parameter exists or not before attempting to access it. And the implementation of the filler function in `fsp_fuse_intf_AddDirInfo` will have no way of knowing whether the file system was compiled with the `FSP_FUSE_READDIR_PLUS` conditional macro enabled or not. [Unless we pass this information into the `fsp_fuse_env`...]

Another solution that has crossed my mind before is to implement both the FUSE 2.8 and 3.0 API's for WinFsp. The FUSE 2.8 API is implemented by the `fsp_fuse_*` symbols and the FUSE 3.0 API could be implemented by `fsp_fuse3_*` symbols. Obviously this is non-trivial and might not help file systems that are not FUSE 3 compatible.

I need to think more about this. As RC2 is now tested and almost ready for release, I do not think such a feature can make it in for a while.

**Comment by netheril96** [comment]
Well, you can introduce a mounting option that always require the `stat` buffer in `fuse_filler_dir` to be fully filled, so that the extra flags can be ignored.

**Comment by billziss-gh** [comment]
It's a bit of hack, but so is my suggestion, so let me give it some thought.

**Comment by billziss-gh** [comment]
Hey, @netheril96. This is not completely related although it has to do with performance.

Are you using `-oFileInfoTimeout=1000` or `-oattr_timeout=1`? This will instruct WinFsp to cache your file system's metadata (for 1 second) thus significantly increasing performance.

You can also set `oFileInfoTimeout=-1`, which will instruct WinFsp to cache file data using the NTOS cache manager. This will make your file system be as fast as a native file system for read/writes (on cached files).

If you do decide to enable full caching, keep in mind that:

- Files in the underlying file system (NTFS) should not be changed by any agent other than your file system or you can hit consistency issues.
- It may be best to disable caching on the underlying file system (NTFS) by opening files with `FILE_FLAG_NO_BUFFERING`. This will avoid double-caching your files. There are some additional requirements to use `FILE_FLAG_NO_BUFFERING`, you have to write files in increments of the underlying allocation unit (I think 4096 on modern NTFS systems, but check with `GetDiskFreeSpaceW`).

For performance testing of using WinFsp with different values of `FileInfoTimeout` see here:

https://github.com/billziss-gh/winfsp/wiki/WinFsp-Performance-Testing

**Comment by netheril96** [comment]
See http://manpages.ubuntu.com/manpages/precise/man8/mount.fuse.8.html.

The `attr_timeout` should default to 1, or at least that is what Ubuntu maintainers think.

**Comment by netheril96** [comment]
And what kind of  options are supported by WinFsp? Is there a list?

**Comment by billziss-gh** [comment]
> The `attr_timeout` should default to 1, or at least that is what Ubuntu maintainers think.

I think the Ubuntu maintainers are correct: https://github.com/libfuse/libfuse/blob/fuse_2_8_bugfix/lib/fuse.c#L3728

The WinFsp-FUSE layer does not currently do this. I seem to recall mulling over this when the code was originally written and eventually deciding against it, because the FUSE `attr_timeout` notion is not exactly `FileInfoTimeout` and because non-FUSE (but WinFsp) file systems get `FileInfoTimeout==0` by default.

In retrospect this may have been a bad decision, because the default option makes file systems unnecessarily slow. Do you have an opinion on the default value? It should be an easy fix.

> And what kind of options are supported by WinFsp? Is there a list?

Running the file system with `-h` will list the WinFsp-specific options. For the full list, the answer is the source code: https://github.com/billziss-gh/winfsp/blob/master/src/dll/fuse/fuse.c#L48

**Comment by billziss-gh** [comment]
> Well, you can introduce a mounting option that always require the stat buffer in fuse_filler_dir to be fully filled, so that the extra flags can be ignored.

BTW, I have thought some more about this. I dislike the idea of allowing the user to specify a `-o ReadDirPlus` option or similar, because the user may then erroneously add the switch for a file system that does not support (our version of) `readdirplus`.

A better alternative is to use the `fuse_conn_info` structure to communicate this information. There is precedent here; OSXFUSE does the same thing for some of its own flags: https://github.com/osxfuse/fuse/blob/support/osxfuse-2/include/fuse_common.h#L155

**Comment by tstordyallison** [comment]
+1 for this. 

I've found in testing that I get lots of gettattr calls along with readdir (that are invariably internal cache hits). As I'm also using fusepy, and I'm constrained somewhat by the GIL, it quickly becomes a bottleneck.

**Comment by tstordyallison** [comment]
Also, unless I'm missing something - the caching 'attr_timeout' setting or NTOS cache manager setting don't apply here, as the (f)getattr calls are generated by the winfsp-*.dll library inside of the ReadDirectory callback.

I tried various attr_timeout settings at first, and then I read the code :)

**Comment by netheril96** [comment]
fuse_conn_info is a good way.

**Comment by billziss-gh** [comment]
@tstordyallison wrote:
> Also, unless I'm missing something - the caching 'attr_timeout' setting or NTOS cache manager setting don't apply here, as the (f)getattr calls are generated by the winfsp-*.dll library inside of the ReadDirectory callback.

You are correct. The FUSE layer does not do any caching of its own. The `FileInfoTimeout` / `attr_timeout` only controls the FSD's caching of file data / metadata. As `readdir` is currently implemented, you will get the extra `(f)getattr` calls (as you correctly observed).

@netheril96 wrote:
> fuse_conn_info is a good way.

I am currently testing RC2 for submission and signing by Microsoft, so this cannot make it in RC2. I promise to look into it (and the default value for `attr_timeout`) right after.

BTW, I also accept patches ;)

**Comment by billziss-gh** [comment]
Commit 712870ddd976db551d80ef8b8d5c5f2637c8af55 adds some WinFsp-FUSE specific capabilities:

    #define FSP_FUSE_CAP_READDIR_PLUS       (1 << 21)   /* file system supports enhanced readdir */
    #define FSP_FUSE_CAP_READ_ONLY          (1 << 22)   /* file system is marked read-only */
    #define FSP_FUSE_CAP_CASE_INSENSITIVE   FUSE_CAP_CASE_INSENSITIVE

This works as follows: when you receive the FUSE `init` call, these bits are set in `conn_info->capable`. If your file system supports any of these capabilities you are supposed to set them in `conn_info->want`.

Here is how I do it in `passthrough-fuse`:

    #if defined(FSP_FUSE_CAP_READDIR_PLUS)
        conn->want |= (conn->capable & FSP_FUSE_CAP_READDIR_PLUS);
    #endif

    #if defined(FSP_FUSE_CAP_CASE_INSENSITIVE)
        conn->want |= (conn->capable & FSP_FUSE_CAP_CASE_INSENSITIVE);
    #endif

This commit is currently on the pvt-fuse-fixes branch. I have tested it locally, but because AppVeyor seems to be having some issues with its Visual Studio license, I have not tested it fully yet.

**Comment by billziss-gh** [comment]
These changes have now been tested and merged into release/1.0. They will appear in RC3.

**Comment by tstordyallison** [comment]
Thanks!

**Comment by benrubson** [comment]
Hi,
Does this patch finally improve performance ?
Does it avoid additional `getattr` on each directory entry ?

I thought that Fuse would have expected `st_ino` to be filled for this to work, but does not seem you compute an inode number in https://github.com/billziss-gh/winfsp/commit/712870ddd976db551d80ef8b8d5c5f2637c8af55 @billziss-gh ?

Thank you very much 👍

**Comment by billziss-gh** [comment]
@benrubson yes, the patch works and makes things much faster. If you find otherwise it is a bug.

Code: [`fsp_fuse_intf_AddDirInfo`](https://github.com/billziss-gh/winfsp/blob/v1.2/src/dll/fuse/fuse_intf.c#L1616) calls [`fsp_fuse_intf_GetFileInfoFunnel `](https://github.com/billziss-gh/winfsp/blob/v1.2/src/dll/fuse/fuse_intf.c#L1656) which will copy [`st_ino`](https://github.com/billziss-gh/winfsp/blob/v1.2/src/dll/fuse/fuse_intf.c#L435).

Notice however that the `st_ino` field is not really used in Windows (at least not in the way it is used on UNIX). So even if you do not set it, things will work just fine.

**Comment by benrubson** [comment]
Thank you @billziss-gh for your answer !

I meant in `readdir()`, in the code you added [to fill the stat structure](https://github.com/billziss-gh/winfsp/commit/712870ddd976db551d80ef8b8d5c5f2637c8af55#diff-49f9770117a964ef371914ded2a1bb4c), you do not compute inode number. So then inode number is 0 in `fsp_fuse_intf_GetFileInfoFunnel`, right ?

Sounds like though that the method to compute the inode number is to use `GetFileInformationByHandle` to get `nFileSizeHigh` & `nFileSizeLow`, so would need to open each file in `readdir()`.

If I'm right above, let's then assume inode is set to 0 because not computed in `readdir()`.
If later in the code the file is stated to get the inode number, where will the inode number come from ?
From the Fuse cache, so 0 ?
From the file itself which will be truly stated ?

In other words, should we loose time opening every file to compute the inode number in `readdir()` if it's then cached and retrieve from the cache ? Or do we simply need to do this in stat / getattr function ?

I'm thinking about EncFS which uses `st_ino` in some of its crypto functions and expects it to be non 0.

Thank you again 👍

**Comment by billziss-gh** [comment]
> I meant in readdir(), in the code you added to fill the stat structure, you do not compute inode number. So then inode number is 0 in fsp_fuse_intf_GetFileInfoFunnel, right ?

Ok, I see now that you meant in the passthrough-fuse file system.

The Windows `FindFirstFile`/`FindNextFile` do not return `IndexNumber`. So you are correct that if it was needed, you would have to do `GetFileInformationByHandle` to get this field. But because Windows does not really use this field (other than in API's like `GetFileInformationByHandle`) I chose not to send it here to simplify the code.

> Sounds like though that the method to compute the inode number is to use GetFileInformationByHandle to get nFileSizeHigh & nFileSizeLow, so would need to open each file in readdir().

You are correct. (I am sure this is a typo but the correct fields are `nFileIndexHigh` and `nFileIndexLow`).

> If I'm right above, let's then assume inode is set to 0 because not computed in readdir().

You are correct. The inode number is 0 in the case of passthrough-fuse during readdir.

> If later in the code the file is stated to get the inode number, where will the inode number come from ?
> From the Fuse cache, so 0 ?
> From the file itself which will be truly stated ?

The directory cache and file attribute cache are separate in the case of Windows. For example, it is actually possible in Windows to query for the size of a file using `FindFirstFile`/`FindNextFile` and get a different size that what you get doing `GetFileInformationByHandle`.

> In other words, should we loose time opening every file to compute the inode number in readdir() if it's then cached and retrieve from the cache ? Or do we simply need to do this in stat / getattr function ?

No need to do it in `readdir`. If it is important for your scenario, do it in `getattr` only. Keep in mind that the OS will not do anything with this value, but it may be important to the app that uses your file system.

(NOTE: passthrough-fuse does not appear to set `st_ino` anywhere. As explained this is fine in Windows.)

**Comment by billziss-gh** [comment]
One additional comment:

If you want to do it in `readdir` you can query the underlying file system using [`NtQueryDirectoryFile`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/nf-ntifs-ntquerydirectoryfile) which is the "system call" behind `FindFirstFileW`/`FindNextFileW`. Query for `FileIdFullDirectoryInformation` and you will get the `IndexNumber` under the field [`FileId`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/ntifs/ns-ntifs-_file_id_full_dir_information).

In the Win32 API the API's used to get the `IndexNumber` are `GetFileInformationByHandle` and `GetFileInformationByHandleEx`. Furthermore the `IndexNumber` is purely informational. So I think it is safe to just return `st_ino` and `IndexNumber` in `getattr` only.

**Comment by benrubson** [comment]
Some last comments/questions to be sure I fully understand :)

If we provide a fully filled stat structure to the Fuse `filler()` function, this function will get the info directly from the structure thus saving an additional stat call. Nice.

I clearly understand `st_ino` is useless to Windows itself.
So no need to fill it into the stat structure given to `filler()`.
However, I thought that Fuse was performing some caching of the stated data, and that for caching to work, `st_ino` and `st_mode` were needed.
Perhaps it's only true with Fuse on Linux, perhaps I also may be wrong...

`st_ino` for sure is important to (for example) Fuse filesystem EncFS.
EncFS will internally get this `st_ino` field `stat()`ing the file (this is what it does it its Linux code).
According to your previous answer, the info needed by `stat()` can't be retrieved from the (cached ?) stat info previously computed by `readdir`.
So in this case, the file will have to be opened by `stat()` to compute `st_ino`, so I think there's no reason to loose time / increase complexity doing this in `readdir`.

Last question (but not least!), what about Cygwin on top of for example Fuse filesystem EncFS.
When we do `ls -l` in Cygwin over EncFS. Or `rsync -a` which needs stat info.
Internally, `readdir` will be called, stat struct will be given to `filler()` for every item in the directory...
Will an additional `getattr` call be fired by Cygwin for every item ? I would say no ?
If we add `-i` to `ls` command (or if `rsync` needs the inode number), where will Cygwin take it from ?
Perhaps then the risk here is to get `0` if `st_ino` has not been computed by `readdir` ?

Quite a few questions, but seems important to clearly understand the background to do things properly, as you see I still have some difficulties to put things into the right order. Then thank you again 👍

**Comment by billziss-gh** [comment]
> However, I thought that Fuse was performing some caching of the stated data, and that for caching to work, st_ino and st_mode were needed.
Perhaps it's only true with Fuse on Linux, perhaps I also may be wrong...

The Windows OS (NTOS) does not provide a specialized directory or file attribute cache. Instead it provides a generic mechanism (called the Cache Manager) that can be used to cache file streams.

Windows file systems and WinFsp use the Cache Manager and other mechanisms to cache file data and metadata. File metadata in our case includes file attributes (`FileInfo` in WinFsp parlance) and directory data (`DirInfo` in WinFsp parlance). Notice that `DirInfo` also contains file attributes on Windows (contrary to POSIX).

Although `DirInfo` is a superset of `FileInfo` the two data structures are cached independently. The `FileInfo` is cached with the file that it belongs to and the DirInfo is cached with the parent directory of the file that it belongs to.

It is possible for these two structures to get out of sync in the cache, which is legal in Windows (although WinFsp does take steps to minimize this). For example, in NTFS when you write a file its size may not be updated immediately if you read its directory entry, but you will see the correct size if you do `GetFileInformationByHandle`.

**In WinFsp these caches are currently keyed by the file path rather than the i-node.** The main reason for this is that in Windows file lookups are path-based and not (ino, name)-based. For example, an OPEN of the path `\comp1\comp2\...\compN` will be passed as is to the file system driver by Windows and will not be broken down to the usual `LOOKUP root, comp1`, `LOOKUP comp1_ino, comp2`, etc.

(This of course also means that I will have a hard time introducing hard links in WinFsp as I would have to change this fundamental data structure and at that point I will be forced to reintroduce the concept of an "inode number" or "file id", etc.)

So Windows does not use `st_ino` (other than for informational purposes in `GetFileInformationByHandle`) and WinFsp does not use `st_ino`. Caching works because it is keyed off the file path. And yes, `RENAME` is a nightmare because of this :)

> EncFS will internally get this st_ino field stat()ing the file (this is what it does it its Linux code).
According to your previous answer, the info needed by stat() can't be retrieved from the (cached ?) stat info previously computed by readdir.

I am not familiar with the EncFS internal workings. Is EncFS a layer over an existing file system? If so, then EncFS can get the needed inode numbers (`IndexNumber`) from the underlying file system (NTFS). NTFS will provide real `IndexNumber`, which can be used as `st_ino` for EncFS purposes. Whether it reports `st_ino==0` or something else to the OS is irrelevant to its operation, I would think.

But it is likely that I misunderstand how EncFS works. I know little about this file system.

> Last question (but not least!), what about Cygwin on top of for example Fuse filesystem EncFS.
When we do ls -l in Cygwin over EncFS. Or rsync -a which needs stat info.
Internally, readdir will be called, stat struct will be given to filler() for every item in the directory...
Will an additional getattr call be fired by Cygwin for every item ? I would say no ?
If we add -i to ls command (or if rsync needs the inode number), where will Cygwin take it from ?
Perhaps then the risk here is to get 0 if st_ino has not been computed by readdir ?

It has been a while since I studied the Cygwin source. So my answers are (semi-educated) guesses.

Regarding `ls -l`. Yes, I fully expect that it does a `readdir` followed by `stat` as in UNIX. However I believe that Cygwin uses `NtQueryDirectoryFile` for `stat` and friends. This has the benefit of going through the same directory cache that WinFsp maintains and that `readdir` just populated. (A brief look to a slightly dated version of Cygwin reveals that `fhandler_base::fstat_by_name` appears to call `NtQueryDirectoryFile`. This lives in file `winsup\cygwin\fhandler_disk_file.cc`.)

Regarding `st_ino==0`. I believe Cygwin checks for this case and conjures an inode number from the file name hash.

**Comment by benrubson** [comment]
@billziss-gh thank you very much for your (very-detailed) answer, much appreciated 👍 
I now understand better, perfect.

EncFS is a userspace filesystem, using the FUSE library.
And so you clearly answered my question, `DirInfo` and `FileInfo` are cached independently, so when EncFS will `stat()` the file (let's say for the first time), if `stat()` uses `GetFileInformationByHandle`, info will be retrieved from `FileInfo`, so we will make an additional disk IO as the cache will be empty. But it `stat()` uses `NtQueryDirectoryFile`, info will be retrieved from `DirInfo`, which will already be populated by `readdir`. Nice, really interesting !

I hope this will help others too :)

Many thanks @billziss-gh for your great support 👍

**Comment by billziss-gh** [comment]
You are welcome :)
